packages/fxa-auth-server/lib/routes/session.js

```diff
@@ -368,20 +370,22 @@ module.exports = function(
           throw error.invalidOrExpiredOtpCode();
         }
 
+        // If a valid code was sent, this verifies the session using the `email-2fa` method.
+        // The assurance level will be ["pwd", "email"] or level 1.
+        await db.verifyTokensWithMethod(sessionToken.id, 'email-2fa');
+
         // We have a matching code! Let's verify the account, session and send the
         // corresponding email and emit metrics.
         if (!account.primaryEmail.isVerified) {
           await signupUtils.verifyAccount(request, account, options);
         } else {
-          const { uid } = sessionToken;
           request.emitMetricsEvent('account.confirmed', { uid });
-          const devices = await request.app.devices;
-          await push.notifyAccountUpdated(uid, devices, 'accountConfirm');
         }
 
-        // If a valid code was sent, this verifies the session using the `email-2fa` method.
-        // The assurance level will be ["pwd", "email"] or level 1.
-        await db.verifyTokensWithMethod(sessionToken.id, 'email-2fa');
+        // To avoid any potential race conditions, the very last thing done is send
+        // a push notifications to connected devices that the account and
+        // session were verified.
+        await push.notifyAccountUpdated(uid, devices, 'accountConfirm');
```

Comment:
With the latest update, I think this could stay in the `else` branch - you don't want to end up sending *both* the `accountVerify` notification and the `accountConfirm` notification in the case where `!account.primaryEmail.isVerified` is true.